### PR TITLE
fix: properly track json memory usage

### PR DESCRIPTION
**Pull request by kostasrim**
There is a problem on how we track memory usage for `JSON` family. Specifically the expression below:

```
return zmalloc_size(u_.json_obj.json_ptr);
```

is wrong and it always returns the same fixed size (16) regardless of the actual memory used by the json object. It's wrong because this returns the size of the object pointed to by `json_ptr` but it does not account the objects it allocates internally (it's internal pointers). For that to happen we would need to track `Json`'s object memory somehow but this is complicated (see in comments).

* add JsonMemTracker
* add logic based on MiMallocResource deltas that calculates json object usage
* add test

Addresses #3661

### CODE REVIEW

File: src/core/compact_object.cc
Comment (kostasrim) [inline src/core/compact_object.cc:1044]: This is also wrong. `ptr->capacity` does not return the actual memory consumed in bytes. So clearly we have a problem. The main issues are:

1. There doesn't seem to exist a function in `Json` that clearly returns the size in bytes used by the object.
2. We can't approximate this easily since `json elements are polymoprhic` and it does not use flat storage internally such that we can use `zmalloc_usable_size()` on its `internal pointer`. 

We could in practice iterate over all of its elements and write a visitor that aggregates the size of all of its elements but that would be extremely slow especially when `info memory` command is used because now we need to iterate over all the json keys in the data store -- obviously a `no no`

And we also don't really control `json` since it's an external library. 

So my idea is to wrap `class MiMemoryResource` and introduce a `MiJsonMemoryResource`. This allocator will use the `same heap` as the `CompactObject::memory_resource` with one difference. It will have a member which will keep track how much memory it allocated/deallocated. Once the json operation finishes, we will reset this member value (such that subsequent operations on different json keys get accurate memory usage) and add/subtract it from a `size` object which will introduce in `JsonWrapper`. That way, when we ask for memory we will return this field without calculating anything and the memory consumption will properly be tracked by the allocator eliminating the need for manual tracking on object that belongs to a library we don't own. 

WDYT? @chakaz @adiholden

File: src/core/compact_object.cc
Comment (kostasrim) [inline src/core/compact_object.cc:700]: Bingo. 

```
2024-09-03T16:48:59.0738486Z 60: F20240903 16:48:59.073283 22070 compact_object.cc:700] Check failed: std::pmr::polymorphic_allocator<char>{memory_resource()} == u_.json_obj.json_ptr->get_allocator() 
2024-09-03T16:48:59.0740855Z 60: *** Check failure stack trace: ***
2024-09-03T16:48:59.0761492Z 60:     @     0x564941768dbb  google::LogMessage::Fail()
2024-09-03T16:48:59.0781948Z 60:     @     0x564941768d02  google::LogMessage::SendToLog()
2024-09-03T16:48:59.0802555Z 60:     @     0x5649417684f7  google::LogMessage::Flush()
2024-09-03T16:48:59.0824658Z 60:     @     0x56494176c20e  google::LogMessageFatal::~LogMessageFatal()
2024-09-03T16:48:59.0869710Z 60:     @     0x5649412c6855  dfly::CompactObj::SetJson()
2024-09-03T16:48:59.0895859Z 60:     @     0x564940be862b  dfly::RdbLoaderBase::OpaqueObjLoader::HandleBlob()
2024-09-03T16:48:59.0930073Z 60:     @     0x564940be4b97  dfly::RdbLoaderBase::OpaqueObjLoader::operator()()
2024-09-03T16:48:59.0972643Z 60:     @     0x564940c0b783  std::__invoke_impl<>()
2024-09-03T16:48:59.1012552Z 60:     @     0x564940c05aaa  std::__invoke<>()
2024-09-03T16:48:59.1059136Z 60:     @     0x564940bffb53  std::__detail::__variant::__gen_vtable_impl<>::__visit_invoke_impl()
2024-09-03T16:48:59.1091480Z 60:     @     0x564940bffb96  std::__detail::__variant::__gen_vtable_impl<>::__do_visit_invoke()
```

`memory_resource was not propagated`.

Comment (kostasrim) [reply src/core/compact_object.cc:700]: From my understanding the allocate used should have been `CompactObject::memory_resource()`. I haven't checked the specific case but based on the semantics on the json family this should be the case. Let me know if I am wrong.

File: src/core/compact_object.cc
Comment (chakaz) [reply src/core/compact_object.cc:1044]: Generally, using a custom memory resource for JSON sounds like a good solution. Indeed it would just call the data heap so all data will be in the same place as it was before.

This however won't play very nicely with the way we compute object memory today, and in many places we want to know the size of a specific object. So I wonder, perhaps we want to have a memory_resource per JSON object? It has the cost of 2 pointers per object, so not sure if it's worth it..

File: src/core/compact_object.cc
Comment (chakaz) [reply src/core/compact_object.cc:700]: If you think this is a problem with jsoncons, I'd try to create a simple standalone example of this happening and report it in their repo.
We've already reported similar issues before, and Daniel (the maintainer) always helped us

File: src/core/compact_object.cc
Comment (kostasrim) [reply src/core/compact_object.cc:1044]: > know the size of a specific object. 
Yep and that why will introduce a single size_t per json object in CompactObj. So the overhead will only be for one element which will keep the size (which is fine because we need a way to track memory for json objects)

**Comment by romange** [comment]
In the worst case we can handle json differently from the other objects by using a separate MiMemoryResource with the same data heap. `MiMemoryResource` will count used memory, so will know the aggregate of JSON objects.

We won't know "memory usage <key>" but this is much smaller problem.

**Comment by kostasrim** [comment]
> In the worst case we can handle json differently from the other objects by using a separate MiMemoryResource with the same data heap. `MiMemoryResource` will count used memory, so will know the aggregate of JSON objects.
> 
> We won't know "memory usage " but this is much smaller problem.

that's the plan and it's almost ready. But I found bugs in their json impl not propagating the allocators properly. I am patching this and I will raise the issue on jsoncons lib

File: src/core/compact_object.h
Comment (kostasrim) [inline src/core/compact_object.h:452]: We don't really need this but I want to keep the DCHECKS. If in the future we decide we need an extra `8` bytes for `JsonConsT`, we can remove this :man_dancing:

File: src/core/json/json_object.cc
Comment (kostasrim) [inline src/core/json/json_object.cc:30]: we are blocked on this until it's addressed by jsoncons

Comment (romange) [reply src/core/json/json_object.cc:30]: so add an issue in the comment and comment out the DCHECK I guess?

File: src/core/compact_object.h
Comment (romange) [inline src/core/compact_object.h:452]: I am not sure why you keep `MiMemoryResource* mr;` inside each CompactObject.
1. JsonType already has allocator in it
2. All json objects will use the same MiMemoryResource instance unless I am missing something

File: src/server/json_family.cc
Comment (romange) [inline src/server/json_family.cc:18]: please do not include `<memory_resource>` directly and in general, can you please disable include what your use rule in your environment? it's too agressive

File: src/core/json/json_object.cc
Comment (kostasrim) [reply src/core/json/json_object.cc:30]: it will cause a memory leak. We will loose this `mr`. Yes maybe we can release it here but then the tracking will be wrong. :/

File: src/core/compact_object.h
Comment (kostasrim) [reply src/core/compact_object.h:452]: I know about this. I keep it because I want to `DCHECK` the hell out of it. It's just breaks with `pmr` in so many levels that even if we rolled in production I would add `LOG(warning)` to warn us that the `pmr` went brrr

**Comment by kostasrim** [comment]
I tested this locally but I am unsure how I can write explicit tests for memory usage that will work on the CI. My assumption here is that different platforms won't have the exact same memory usage but maybe I am wrong. WDYT @romange @chakaz

File: src/core/compact_object.cc
Comment (kostasrim) [inline src/core/compact_object.cc:56]: maybe worth making it thread local

Comment (chakaz) [inline src/core/compact_object.cc:56]:
```suggestion
  static bool is_flat = absl::GetFlag(FLAGS_experimental_flat_json);
```

Comment (chakaz) [reply src/core/compact_object.cc:56]: I'd do that, why not?

File: src/core/compact_object.cc
Comment (chakaz) [inline src/core/compact_object.cc:728]: Why not accept `size` (call it `delta`) as some signed type (`int64_t` or whatever), and then just do `u_.json_obj.cons.bytes_used += delta;`?
Why all the `if`s and extra parameter?

Comment (chakaz) [reply src/core/compact_object.cc:728]: Also specifically this first `if` is unneeded right?

File: src/core/compact_object.h
Comment (romange) [inline src/core/compact_object.h:453]: why not reduce it to uint32_t and keep the encoding inside the object?

File: src/core/compact_object.h
Comment (romange) [inline src/core/compact_object.h:453]: why not keep both fields common for both encodings?

File: src/server/json_family.cc
Comment (romange) [inline src/server/json_family.cc:64]: please refrain from using default arguments

File: src/server/json_family.cc
Comment (romange) [inline src/server/json_family.cc:67]: we now call Find 3 times more times?

Comment (kostasrim) [reply src/server/json_family.cc:67]: 2 more times in the worst case (when the json does not exist). Otherwise it's one.

File: src/core/compact_object.h
Comment (kostasrim) [reply src/core/compact_object.h:453]: IMO len is how many characters. The other one is how many bytes. I find it a little clearer :shrug:

File: src/server/json_family.cc
Comment (kostasrim) [reply src/server/json_family.cc:64]: sure will remove

File: src/core/compact_object.h
Comment (kostasrim) [reply src/core/compact_object.h:453]: uint_32 limits to 4gb. It's extreme to expect any object would take more than that but you never know. As for enconding, I don't think it makes much sense to store this. It's fixed once it's set and that's also one of the reasons I refactored it

File: src/core/compact_object.cc
Comment (kostasrim) [reply src/core/compact_object.cc:728]: yes I agree we don't need the first case. Furthermore, I was afraid of overflowing/underflowing but we do the exact same thing as you suggest in `AccountObjectMemory` so I guess it should not be an issue. Delta sounds simpler also. I will apply this.

File: src/core/compact_object.h
Comment (romange) [reply src/core/compact_object.h:453]: The reason for storing it per object is that I expect we would want a heuristic at some point for encoding, managed per object basis.  like we do with hash maps, sets etc. flat is awful for mutable operations, for example.  we do not have to do it now though.

File: src/server/json_family.cc
Comment (romange) [reply src/server/json_family.cc:67]: is it possible to use a similar pattern with already existing iterator? i.e. avoid additional finds?

File: src/core/compact_object.h
Comment (kostasrim) [reply src/core/compact_object.h:453]: I wanted to pack this in the beginning and in fact I already mentioned to @adiholden that if in the future we wanted more space on the json we store within a compcat object we could get some. However I thought that the size_t suffices for now and until we actually need this there is no reason to take the bitfield route.

Comment (kostasrim) [reply src/core/compact_object.h:453]: also if we ever want a heuristic, how many bits would that require? 4-8 ? For that use case we will just make the bytes_used (and the length for flat json) a bitfield and mask the last `n` bits we want to use for enconding. It would be a simple change that would server our need.

File: src/server/json_family.cc
Comment (chakaz) [reply src/server/json_family.cc:64]: reminder :)

File: src/server/json_family.cc
Comment (chakaz) [reply src/server/json_family.cc:67]: Can we modify `JsonMemTracker` to use the compact object directly (i.e. not to call `Find()`) and modify `UpdateEntry()` to use it?
And even if I missed some call that it won't work for, we can code around that instead, because it does seem like all (the majority?) of calls use it

File: src/core/compact_object.cc
Comment (kostasrim) [reply src/core/compact_object.cc:55]: yes you are right. For some reason I always forget about the order of initialization for globals. Same function as a I had before just with a thread_local within it :)

File: src/server/json_family.cc
Comment (chakaz) [inline src/server/json_family.cc:41]: I'd expose the function instead of using the flag here directly

**Review by chakaz — COMMENTED** [review]
Generally LG! Please see some last nits

**Review by chakaz — APPROVED** [review]

**Pull request merged**
